### PR TITLE
feat: Add BatchEnforce() API

### PR DESCRIPTION
**Pull request by RobotHuang**
fix: #617

### CODE REVIEW

**Comment by hsluoyz** [comment]
@RobotHuang plz fix:

image

File: enforcer.go [hunk 1]
```diff
@@ -610,6 +611,34 @@ func (e *Enforcer) EnforceExWithMatcher(matcher string, rvals ...interface{}) (b
 	return result, explain, err
 }
 
+// BatchEnforce enforce in batches
+func (e *Enforcer) BatchEnforce(requests [][]string) (map[string]bool, error) {
```
Comment (nodece) [inline enforcer.go:615]: requests should be `[][]interface`.

File: enforcer.go [hunk 2]
```diff
@@ -610,6 +611,34 @@ func (e *Enforcer) EnforceExWithMatcher(matcher string, rvals ...interface{}) (b
...
+	defaultSep := ", "
+	resultMap := make(map[string]bool)
+	for _, request := range requests {
+		result, err := e.enforce("", nil, request[0], request[1], request[2])
```
Comment (nodece) [inline enforcer.go:619]: you have to pass all the requests to `enforce`.

File: enforcer.go [hunk 3]
```diff
@@ -610,6 +611,34 @@ func (e *Enforcer) EnforceExWithMatcher(matcher string, rvals ...interface{}) (b
...
+		if err != nil {
+			return nil, err
+		}
+		resultMap[strings.Join(request, defaultSep)] = result
+	}
+	return resultMap, nil
+}
+
+// BatchEnforceWithMatcher enforce with matcher in batches
+func (e *Enforcer) BatchEnforceWithMatcher(matcher string, requests [][]string) (map[string]bool, error) {
```
Comment (nodece) [inline enforcer.go:629]: requests should be `[][]interface{}`.

File: enforcer.go [hunk 4]
```diff
@@ -610,6 +611,34 @@ func (e *Enforcer) EnforceExWithMatcher(matcher string, rvals ...interface{}) (b
...
+	defaultSep := ", "
+	resultMap := make(map[string]bool)
+	for _, request := range requests {
+		result, err := e.enforce(matcher, nil, request[0], request[1], request[2])
```
Comment (nodece) [inline enforcer.go:633]: you have to pass all the requests to `enforce`.

File: enforcer.go (hunk 1 above)
Comment (RobotHuang) [reply enforcer.go:615]: If we change to `[][]interface{}`, we couldn't pass `[][]string{}` to this function.

File: enforcer_test.go [hunk 5]
```diff
@@ -479,3 +480,19 @@ func TestEnforceExLog(t *testing.T) {
...
+func testBatchEnforce(t *testing.T, e *Enforcer, requests [][]interface{}, results map[string]bool) {
+	t.Helper()
+	if myRes, _ := e.BatchEnforce(requests); !reflect.DeepEqual(myRes, results) {
+		t.Errorf("%v supposed to be %v", myRes, results)
+	}
+
+}
+func TestBatchEnforce(t *testing.T) {
+	e, _ := NewEnforcer("examples/basic_model.conf", "examples/basic_policy.csv")
+	results := make(map[string]bool)
```
Comment (hsluoyz) [inline enforcer_test.go:493]: 4 lines to 1 line

File: enforcer.go [hunk 6]
```diff
@@ -610,6 +611,50 @@ func (e *Enforcer) EnforceExWithMatcher(matcher string, rvals ...interface{}) (b
...
+		for _, item := range request {
+			value, ok := item.(string)
+			if !ok {
+				return nil, errors.New("please make sure that every value in request array is string")
+			}
+			requestStr = append(requestStr, value)
+		}
+		resultMap[strings.Join(requestStr, defaultSep)] = result
+	}
+	return resultMap, nil
```
Comment (hsluoyz) [inline enforcer.go:633]: Return slice

File: enforcer.go [hunk 7]
```diff
@@ -610,6 +611,50 @@ func (e *Enforcer) EnforceExWithMatcher(matcher string, rvals ...interface{}) (b
...
+	defaultSep := ", "
+	resultMap := make(map[string]bool)
+	for _, request := range requests {
+		result, err := e.enforce("", nil, request...)
+		if err != nil {
+			return nil, err
+		}
+		var requestStr []string
+		for _, item := range request {
+			value, ok := item.(string)
```
Comment (hsluoyz) [inline enforcer.go:625]: `e.enforce()` supports struct arg, see ABAC. It doesn't have to be string.

File: enforcer_test.go [hunk 8]
```diff
@@ -479,3 +480,19 @@ func TestEnforceExLog(t *testing.T) {
 	testEnforceEx(t, e, "bob", "data2", "read", []string{})
 	testEnforceEx(t, e, "bob", "data2", "write", []string{"bob", "data2", "write"})
 }
+
+func testBatchEnforce(t *testing.T, e *Enforcer, requests [][]interface{}, results map[string]bool) {
+	t.Helper()
+	if myRes, _ := e.BatchEnforce(requests); !reflect.DeepEqual(myRes, results) {
```
Comment (hsluoyz) [inline enforcer_test.go:486]: Use the existing `util.Array2DEquals()` instead of reflection.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: enforcer.go [hunk 9]
```diff
@@ -17,7 +17,6 @@ package casbin
 import (
 	"errors"
 	"fmt"
-
```
Comment (hsluoyz) [inline enforcer.go:20]: don't remove new line.

File: enforcer_test.go [hunk 10]
```diff
@@ -479,3 +479,21 @@ func TestEnforceExLog(t *testing.T) {
...
+	if len(myRes) != len(results) {
+		t.Errorf("%v supposed to be %v", myRes, results)
+	}
+	for i, v := range myRes {
+		if v != results[i] {
+			t.Errorf("%v supposed to be %v", myRes, results)
+		}
+	}
+}
+func TestBatchEnforce(t *testing.T) {
```
Comment (hsluoyz) [inline enforcer_test.go:495]: new line.

File: enforcer_test.go [hunk 11]
```diff
@@ -479,3 +479,21 @@ func TestEnforceExLog(t *testing.T) {
...
+	e, _ := NewEnforcer("examples/basic_model.conf", "examples/basic_policy.csv")
+	results := []bool{true, true, false}
+	testBatchEnforce(t, e, [][]interface{}{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"jack", "data3", "read"}}, results)
+}
```
Comment (hsluoyz) [inline enforcer_test.go:499]: new line at end.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Comment by hsluoyz** [comment]
lgtm

**Pull request merged**